[testing_tool] allow TOOL parameter to contain an arbitrary command

fbrausse opened the pull request:
Currently, the `testing_tool.py` requires its TOOL parameter to be a an absolute path to the executable and nothing else, since it is used directly for `Pexec` **after** changing the current working directory. This patch relaxes these requirements and enables
- TOOL to be a string interpreted by the shell (via the `shlex` package in Python's stdlib), i.e., split at (non-escaped/non-quoted) spaces, etc. and the first element of it should be a path (relative or absolute) to the tool. This allows running esbmc for example with default parameters not part of the specific test: `./testing_tool.py --tool '../build/src/esbmc/esbmc --yices' [...]`
- shell-escaping also in the `test_args` string (i.e. the 3rd line in `test.desc`), whereas before it was split at any space character.

fbrausse committed:
[regression/testing_tool] allow TOOL parameter to contain spaces and a relative path

fbrausse commented:
PS: This might break scripts which pass a path containing spaces as `TOOL` parameter. I am not aware whether such configurations do exist, so please regard this PR as an RFC.

rafaelsamenezes reviewed (changes requested):
The code and change itself LGTM.

Could you add a unit test for the a tool with arguments?

Something similar to: https://github.com/esbmc/esbmc/blob/master/regression/testing_tool_test.py#L127-L131

fbrausse committed:
[testing-tool] avoid any chdir before invoking TOOL

This patch makes the testing_tool oblivious to the current working directory.

Generalizes the previous patch by not touching the TOOL parameter in any way
except splitting the string according to shell rules. This has the effect that
arbitrary runtime environments for esbmc can now be specified using the TOOL
parameter, e.g.,

  ../regression/testing_tool.py --tool
    "qemu-aarch64 -L /usr/aarch64-linux-gnu -E [...]
       ../build/src/esbmc/esbmc --fixedbv" [...]

The path to the regression suite is now relative to that of testing_tool.py,
which is determined at runtime. The file paths constructed for the command
line are now generated by prefixing each with the regression suite's path.
A parameter from the 3rd line of test.desc is considered to be a file path
depending on whether it exists when prefixing it by the regression suite's
path.

This has the additional effect that the current working directory no longer
influences the way the regression suites are found.

fbrausse committed:
[testing-tool-test] add test cases for TOOL being a list of parameters

fbrausse committed:
[testing_tool] fix for running --file mode

fbrausse commented:
From the failing Windows regression test:
> EXPECTED TO FIND: ['file main.c line 7 function main$']
> PROGRAM OUTPUT
> ESBMC version 6.8.0 64-bit x86_64 windows
> file D:/a/esbmc/esbmc/regression/esbmc\github_381\main.c: Parsing
> [...]
>  file esbmc\github_381\main.c line 7 function main

The forward slashes appear to come from argv[0] in the invocation of `testing_tool.py`, while it itself inserts the backslashes (due to `os.path.join`). As it seems, esbmc does not expect backslashes in paths.

The following patch should fix that problem, however, an alternative is to keep the entire path for `locationt::get_file()`, which would resolve ambiguous messages for multiple input files with the same name. Which behaviour is preferrable here?

```diff
diff --git a/src/clang-c-frontend/clang_c_convert.cpp b/src/clang-c-frontend/clang_c_convert.cpp
index 21cc54d34..456e65624 100644
--- a/src/clang-c-frontend/clang_c_convert.cpp
+++ b/src/clang-c-frontend/clang_c_convert.cpp
@@ -12,6 +12,8 @@
 #include <llvm/Support/raw_os_ostream.h>
 #pragma GCC diagnostic pop
 
+#include <boost/filesystem.hpp>
+
 #include <clang-c-frontend/padding.h>
 #include <clang-c-frontend/clang_c_convert.h>
 #include <clang-c-frontend/typecast.h>
@@ -2944,10 +2946,7 @@ std::string clang_c_convertert::get_modulename_from_path(std::string path)
 
 std::string clang_c_convertert::get_filename_from_path(std::string path)
 {
-  if(path.find_last_of('/') != std::string::npos)
-    return path.substr(path.find_last_of('/') + 1);
-
-  return path;
+  return boost::filesystem::path(path).filename().string();
 }
 
 symbolt *clang_c_convertert::move_symbol_to_context(symbolt &symbol)
```

fbrausse commented:
I guess, what I'm asking is why c54839eadaa2256d8713a2476e6d8155b3f9ffb9 has been introduced and whether it is OK to revert it, since it makes the above test-case fail on windows when `esmbc` is not run within its associated directory.

fbrausse committed:
[regression] remove assumption that file output is a non-prefixed path in pattern

fbrausse commented:
As stated in https://github.com/esbmc/esbmc/pull/564/commits/87965850a07d4357ed52225cb3d78e398a4c15c8 the `TOOL` parameter can now contain arbitrary commands and `testing_tool.py` won't chdir() anywhere. This provides flexibility to not just run esbmc from any path, but also to run it under some emulation or other changed environments.

rafaelsamenezes reviewed (changes requested):
LGTM

I am not sure about what is failing on macOS though. Let's rerun the CI

lucasccordeiro reviewed (approved):
LGTM.
